compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala

```diff
@@ -651,11 +651,20 @@ trait ConstraintHandling {
     def widenOr(tp: Type) =
       if widenUnions then
         val tpw = tp.widenUnion
-        if (tpw ne tp) && !tpw.isTransparent() && (tpw <:< bound) then tpw else tp
+        if tpw ne tp then
+          if tpw.isTransparent() then
+            // Now also widen singletons of soft unions. Before these were skipped
+            // since we widenUnion on soft unions is independent of whether singletons
+            // are widened or not. This avoids an expensive subtype check in widenSingle,
+            // see 19907_*.scala for test cases.
+            tp.widenSingletons()
```

Comment:
I'm wondering if this shouldn't be:
```suggestion
            val tpw1 = tp.widenSingletons()
            if tpw1 <:< bound then tpw1 else tp
```
since this widening too could get out of bounds.
(this could also be implemented by adding a parameter `skipSoftUnions` to `widenSingle` and call `widenSingle(tp, skipSoftUnions = false)`)